File: ObjcExample/ObjcExample/Demo/GKPageScrollView/Custom/GKCustomSegmentedView.m
//
//  GKCustomSegmentedView.m
//  ObjcExample
//
//  Created by QuintGao on 2025/3/18.
//

#import "GKCustomSegmentedView.h"

@interface GKCustomSegmentedView()

@end

@implementation GKCustomSegmentedView

- (instancetype)initWithFrame:(CGRect)frame {
    if (self = [super initWithFrame:frame]) {
        [self initUI];
    }
    return self;
}

- (void)initUI {
    self.backgroundColor = UIColor.whiteColor;
    
    NSArray *titles = @[@"作品", @"收藏", @"喜欢"];
    [titles enumerateObjectsUsingBlock:^(id  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        UIButton *btn = [[UIButton alloc] init];
        [btn setTitle:obj forState:UIControlStateNormal];
        [btn setTitleColor:UIColor.blackColor forState:UIControlStateNormal];
        btn.titleLabel.font = [UIFont systemFontOfSize:16];
        btn.tag = idx;
        [btn addTarget:self action:@selector(btnClick:) forControlEvents:UIControlEventTouchUpInside];
        [self addSubview:btn];
    }];
}

- (void)layoutSubviews {
    [super layoutSubviews];
    
    CGFloat width = self.bounds.size.width / 3;
    CGFloat height = self.bounds.size.height;
    
    [self.subviews enumerateObjectsUsingBlock:^(__kindof UIView * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        obj.frame = CGRectMake(idx * width, 0, width, height);
    }];
}

- (void)btnClick:(UIButton *)sender {
    NSInteger index = sender.tag;
    
    [self.subviews enumerateObjectsUsingBlock:^(__kindof UIButton * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        if (idx == index) {
            [obj setTitleColor:UIColor.redColor forState:UIControlStateNormal];
        } else {
            [obj setTitleColor:UIColor.blackColor forState:UIControlStateNormal];
        }
    }];
    
    if ([self.delegate respondsToSelector:@selector(didClickSelectAtIndex:)]) {
        [self.delegate didClickSelectAtIndex:index];
    }
}

- (void)scrollSelectWithIndex:(NSInteger)index {
    [self.subviews enumerateObjectsUsingBlock:^(__kindof UIButton * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        if (idx == index) {
            [obj setTitleColor:UIColor.redColor forState:UIControlStateNormal];
        } else {
            [obj setTitleColor:UIColor.blackColor forState:UIControlStateNormal];
        }
    }];
}

@end

File: SwiftExample/SwiftExample/Demo/GKPageScrollView/Custom/GKCustomSegmentedView.swift
//
//  GKCustomSegmentedView.swift
//  SwiftExample
//
//  Created by QuintGao on 2025/3/18.
//

import UIKit

protocol GKCustomSegmentedViewDelegate: NSObjectProtocol {
    func didClickSelect(at index: Int)
}

class GKCustomSegmentedView: UIView {
    
    public weak var delegate: GKCustomSegmentedViewDelegate?

    override init(frame: CGRect) {
        super.init(frame: frame)
        initUI()
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
        initUI()
    }
    
    func initUI() {
        self.backgroundColor = UIColor.white
        let titles = ["作品", "收藏", "喜欢"]
        titles.enumerated().forEach {
            let btn = UIButton()
            btn.setTitle($0.element, for: .normal)
            btn.setTitleColor(.black, for: .normal)
            btn.titleLabel?.font = .systemFont(ofSize: 16)
            btn.tag = $0.offset
            btn.addTarget(self, action: #selector(btnClick), for: .touchUpInside)
            addSubview(btn)
        }
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        
        let width = bounds.width / 3
        let height = bounds.height
        subviews.enumerated().forEach {
            $0.element.frame = CGRect(x: CGFloat($0.offset) * width, y: 0, width: width, height: height)
        }
    }
    
    @objc func btnClick(sender: UIButton) {
        let index = sender.tag
        subviews.enumerated().forEach {
            let btn = $0.element as! UIButton
            if $0.offset == index {
                btn.setTitleColor(.red, for: .normal)
            } else {
                btn.setTitleColor(.black, for: .normal)
            }
        }
        delegate?.didClickSelect(at: index)
    }
    
    public func scrollSelect(with index: Int) {
        subviews.enumerated().forEach {
            let btn = $0.element as! UIButton
            if $0.offset == index {
                btn.setTitleColor(.red, for: .normal)
            } else {
                btn.setTitleColor(.black, for: .normal)
            }
        }
    }
}
